File: src/pipeline/detectors/shell_cannot_open.py
"""
Detector for shell errors when a file cannot be opened.
"""

import re
import typing as T
from src.pipeline.detectors.base import Detector
from src.pipeline.models import ErrorClue


class ShellCannotOpenDetector(Detector):
    """
    Detect shell errors when a file cannot be opened.

    Matches patterns like:
    - sh: 0: cannot open makeoptions: No such file
    - /bin/sh: cannot open file: No such file
    """

    PATTERNS = {
        "missing_file": r"sh:\s*\d+:\s*cannot open\s+(?P<file_path>[^\s:]+):\s*No such file",
    }

    EXAMPLES = [
        (
            "sh: 0: cannot open makeoptions: No such file",
            {
                "clue_type": "missing_file",
                "confidence": 1.0,
                "context": {"file_path": "makeoptions"},
            },
        ),
    ]

File: src/boil/detectors/shell_cannot_open.go
// Detector for shell "cannot open" errors.
package detectors

// ShellCannotOpenDetector detects shell "cannot open" errors
type ShellCannotOpenDetector struct {
	*BaseDetector
}

func NewShellCannotOpenDetector() (*ShellCannotOpenDetector, error) {
	base, err := NewBaseDetector(
		"ShellCannotOpenDetector",
		100,
		map[string]string{
			"missing_file": `sh:\s*\d+:\s*cannot open\s+(?P<file_path>[^\s:]+):\s*No such file`,
		},
		[]DetectorExample{
			{
				Name:     "shell_cannot_open",
				Input:    "sh: 1: cannot open config.sh: No such file",
				ClueType: "missing_file",
				Context:  map[string]string{"file_path": "config.sh"},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &ShellCannotOpenDetector{base}, nil
}
